src/data/ordmap/ordnode.lean

```diff
@@ -0,0 +1,1152 @@
+| nil i := nil
+| t@(node _ l x r) i :=
+  if i = 0 then t else
+  match nat.psub' i (size l) with
+  | none         := link (drop_aux l i) x r
+  | some 0       := insert_min x r
+  | some (j + 1) := drop_aux r j
+  end
+
+/-- O(log n). Remove the first `i` elements of the set, counted from the left.
+
+     drop 2 {a, b, c, d} = {c, d}
+     drop 5 {a, b, c, d} = ∅ -/
+def drop (i : ℕ) (t : ordnode α) : ordnode α :=
+if size t ≤ i then nil else drop_aux t i
+
+/-- Auxiliary definition for `split_at`. (Can also be used in lieu of `split_at` if you know the
+index is within the range of the data structure.)
+
+    split_at_aux {a, b, c, d} 2 = ({a, b}, {c, d})
+    split_at_aux {a, b, c, d} 5 = ({a, b, c, d}, ∅) -/
+def split_at_aux : ordnode α → ℕ → ordnode α × ordnode α
+| nil i := (nil, nil)
+| t@(node _ l x r) i :=
+  if i = 0 then (nil, t) else
+  match nat.psub' i (size l) with
+  | none         := let (l₁, l₂) := split_at_aux l i in (l₁, link l₂ x r)
+  | some 0       := (glue l r, insert_min x r)
+  | some (j + 1) := let (r₁, r₂) := split_at_aux r j in (link l x r₁, r₂)
+  end
+
+/-- O(log n). Split a set at the `i`th element, getting the first `i` and everything else.
+
+     split_at 2 {a, b, c, d} = ({a, b}, {c, d})
+     split_at 5 {a, b, c, d} = ({a, b, c, d}, ∅) -/
+def split_at (i : ℕ) (t : ordnode α) : ordnode α × ordnode α :=
+if size t ≤ i then (t, nil) else split_at_aux t i
+
+/-- O(log n). Get an initial segment of the set that satisfies the predicate `p`.
+`p` is required to be antitone, that is, `x < y → p y → p x`.
+
+    take_while (λ x, x < 4) {1, 2, 3, 4, 5} = {1, 2, 3}
+    take_while (λ x, x > 4) {1, 2, 3, 4, 5} = precondition violation -/
+def take_while (p : α → Prop) [decidable_pred p] : ordnode α → ordnode α
+| nil := nil
+| (node _ l x r) := if p x then link l x (take_while r) else take_while l
+
+/-- O(log n). Remove an initial segment of the set that satisfies the predicate `p`.
+`p` is required to be antitone, that is, `x < y → p y → p x`.
+
+    drop_while (λ x, x < 4) {1, 2, 3, 4, 5} = {4, 5}
+    drop_while (λ x, x > 4) {1, 2, 3, 4, 5} = precondition violation -/
+def drop_while (p : α → Prop) [decidable_pred p] : ordnode α → ordnode α
+| nil := nil
+| (node _ l x r) := if p x then drop_while r else link (drop_while l) x r
+
+/-- O(log n). Split the set into those satisfying and not satisfying the predicate `p`.
+`p` is required to be antitone, that is, `x < y → p y → p x`.
+
+    span (λ x, x < 4) {1, 2, 3, 4, 5} = ({1, 2, 3}, {4, 5})
+    span (λ x, x > 4) {1, 2, 3, 4, 5} = precondition violation -/
+def span (p : α → Prop) [decidable_pred p] : ordnode α → ordnode α × ordnode α
+| nil := (nil, nil)
+| (node _ l x r) :=
+  if p x then let (r₁, r₂) := span r in (link l x r₁, r₂)
+  else        let (l₁, l₂) := span l in (l₁, link l₂ x r)
+
+/-- Auxiliary definition for `of_asc_list`. -/
+def of_asc_list_aux₁ : ∀ l : list α, ℕ → ordnode α × {l' : list α // l'.length ≤ l.length}
```

Comment:
Do these compute okay? This isn't always the case for definitions by well founded recursion.

Reply:
They are written deliberately with well founded recursion because this generates better VM code (IIRC it compiles it to a "naive" / unbounded recursion pattern, while using `rec` produces calls to the recursor). They don't compute well in the kernel though, which is probably what you are thinking of. A similar issue comes up with `prime.decidable` vs `prime.decidable_1`; my thinking is that this data structure is most likely to be used in pure lean code intended for `#eval`, so it is better to prioritize VM evaluation.

Reply:
Yep, that's what I was thinking of. Maybe a short note in the doc string would help here.